zebra-network/src/config.rs

```diff
@@ -176,3 +184,86 @@ impl Default for Config {
         }
     }
 }
+
+impl<'de> Deserialize<'de> for Config {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(deny_unknown_fields, default)]
+        struct DConfig {
+            listen_addr: String,
+            network: Network,
+            initial_mainnet_peers: HashSet<String>,
+            initial_testnet_peers: HashSet<String>,
+            peerset_initial_target_size: usize,
+            #[serde(alias = "new_peer_interval")]
+            crawl_new_peer_interval: Duration,
+        }
+
+        impl Default for DConfig {
+            fn default() -> Self {
+                let config = Config::default();
+                Self {
+                    listen_addr: config.listen_addr.to_string(),
+                    network: config.network,
+                    initial_mainnet_peers: config.initial_mainnet_peers,
+                    initial_testnet_peers: config.initial_testnet_peers,
+                    peerset_initial_target_size: config.peerset_initial_target_size,
+                    crawl_new_peer_interval: config.crawl_new_peer_interval,
+                }
+            }
+        }
+
+        let config = DConfig::deserialize(deserializer)?;
+        let listen_addr = match config.listen_addr.parse::<SocketAddr>() {
+            Ok(socket) => Ok(socket),
+            Err(_) => match config.listen_addr.parse::<IpAddr>() {
+                Ok(ip) => {
+                    let port = match config.network {
+                        Network::Mainnet => 8233,
+                        Network::Testnet => 18233,
+                    };
+                    Ok(SocketAddr::new(ip, port))
+                }
+                Err(err) => Err(de::Error::custom(err)),
```

Comment:
Do we need to mention the accepted formats as part of this error?

Reply:
I think it's ok to leave the error as it. Both `IpAddr` and `SocketAddr` have same error [AddrParseError](https://doc.rust-lang.org/nightly/std/net/struct.AddrParseError.html).

Reply:
I think users will find this error confusing:
> "invalid IP address syntax"
https://doc.rust-lang.org/nightly/src/std/net/parser.rs.html#360

Can we do something like:
```suggestion
                Err(err) => Err(de::Error::custom(format!("{}; Hint: addresses can be a IPv4, IPv6 (with brackets), or a DNS name, the port is optional", err))),
```